src/main/storm/mesos/MesosNimbus.java

```diff
@@ -114,94 +120,228 @@ public String getHostName(Map<String, SupervisorDetails> map, String nodeId) {
-
-        MesosSchedulerDriver driver = createMesosDriver();
-        
-        driver.start();
-    
-        LOG.info("Waiting for scheduler driver to register MesosNimbus with mesos-master and complete initialization...");
-      
-        _scheduler.waitUntilRegistered();
-      
-        LOG.info("Scheduler registration and initialization complete...");
- 
+      initialize(conf, localDir);
+
+      MesosSchedulerDriver driver = createMesosDriver();
+
+      driver.start();
+
+      LOG.info("Waiting for scheduler driver to register MesosNimbus with mesos-master and complete initialization...");
+
+      _scheduler.waitUntilRegistered();
+
+      LOG.info("Scheduler registration and initialization complete...");
+
     } catch (Exception e) {
-        LOG.error("Failed to prepare scheduler ", e);  
+      LOG.error("Failed to prepare scheduler ", e);
       throw new RuntimeException(e);
     }
   }
+
   @SuppressWarnings("unchecked")
   protected void initialize(Map conf, String localDir) throws Exception {
-      _conf  = conf;
-      _state = new LocalState(localDir);
-      _allowedHosts = listIntoSet((List<String>) conf.get(CONF_MESOS_ALLOWED_HOSTS));
-      _disallowedHosts = listIntoSet((List<String>) conf.get(CONF_MESOS_DISALLOWED_HOSTS));
-      _scheduler = new NimbusScheduler();
-      createLocalServerPort();
-      setupHttpServer();
+    _conf = new HashMap();
+    _conf.putAll(conf);
+
+    _state = new LocalState(localDir);
+    _allowedHosts = listIntoSet((List<String>) conf.get(CONF_MESOS_ALLOWED_HOSTS));
+    _disallowedHosts = listIntoSet((List<String>) conf.get(CONF_MESOS_DISALLOWED_HOSTS));
+    Boolean preferReservedResources = (Boolean) conf.get(CONF_MESOS_PREFER_RESERVED_RESOURCES);
+    if (preferReservedResources != null) {
+      _preferReservedResources = preferReservedResources;
+    }
+    _container = Optional.fromNullable((String) conf.get(CONF_MESOS_CONTAINER_DOCKER_IMAGE));
+    _scheduler = new NimbusScheduler(this);
+    createLocalServerPort();
+    setupHttpServer();
+
+    _conf.put(Config.NIMBUS_HOST, _configUrl.getHost());
+    // Generate YAML to be served up to clients
+    Path generatedConfPath = Paths.get("generated-conf");
+    if (!generatedConfPath.toFile().exists() && !generatedConfPath.toFile().mkdirs()) {
+      throw new RuntimeException("Couldn't greate generated-conf dir");
+    }
+    File generatedConf = Paths.get(generatedConfPath.toString(), "storm.yaml").toFile();
+    Yaml yaml = new Yaml();
+    FileWriter writer = new FileWriter(generatedConf);
+    yaml.dump(_conf, writer);
+  }
+
+
+  public void doRegistration(final SchedulerDriver driver, Protos.FrameworkID id) {
+    _driver = driver;
+    try {
+      _state.put(FRAMEWORK_ID, id.getValue());
+    } catch (IOException e) {
+      LOG.error("Halting process...", e);
+      Runtime.getRuntime().halt(1);
+    }
+    Number filterSeconds = Optional.fromNullable((Number) _conf.get(CONF_MESOS_OFFER_FILTER_SECONDS)).or(120);
```

Comment:
Thought provoking response... let me ensure we are on the same page about what filtering means, when it's being applied, and what you are trying to accomplish (I'm sorry about the lack of brevity here, I've spent a bunch of time editing this down, it's a complex subject).

Here is my understanding of the purpose of `Filters.refuse_seconds` (attained after sitting in a zen-like pose and contemplating the deepest secrets of the universe):

* the functionality controls how fast a framework might get a "refused" resource back from the mesos-master, after having chosen not to use it.  And the choice to not use (thus "refuse") a resource is made after having been offered that resource, and the choice is made by either:
 1. a call to `declineOffer()` which declines all resources in the referenced Offer, OR
 2. a call to `launchTasks()` where the launched task(s) don't use *all* of the resources in the chosen Offer(s), thus "refusing" the unused resources.
* even without your changes we are *already* filtering in both cases, since based on [this comment in the mesos.proto file](https://github.com/apache/mesos/blob/0.25.0/include/mesos/mesos.proto#L1166-L1171), the behavior applies to all "unused resources" with a 5 second timeout, whether or not you have chosen to pass Filters explicitly in your call to `declineOffer()` or `launchTasks()`

So my *current* understanding is that we are "filtering" by default even if we are not explicitly choosing to configure the time that we "filter" for.  It's just that the default is for 5 seconds, and you are leaving that alone for `launchTasks()`, and instead setting it to something much larger (120 seconds) for `declineOffer()`.

So now let me reconcile this understanding with your changes and statement about "only filter when we know there are no more workers to launch":

* The default of your new config knob being 120 seconds (way bigger than the 5 second default) implies that when you say "we don't want to start filtering offers until we know there are no more workers to launch", you really mean: "we don't want to filter offers *for a really long time* unless we know there are no more workers to launch".  I conclude that because we *are* filtering for 5 seconds even in the case of `launchTasks()`, so we *are* filtering offers even today whenever we call `declineOffer()` & `launchTasks()`
* Here's my understanding of your desire: you want to back off of receiving any resources back from previously received offers when those offers haven't been used for the TTL of waiting through the RotatingMap.   And *if* we end up needing offers back we will call `reviveOffers()` and then in the next round-trip through the Nimbus scheduling we'll get the resources back.  This seems to be a nice solution to the "greedy hoarding" behavior of the MesosNimbus. Pretty cool.  This feature might obviate the need to directly implement the "only hoard when topologies are pending" idea.

I guess this is all a very long-winded way of saying "it makes sense to do this".  With a couple caveats:

1. We should be strict in our phrasing when discussing this feature.  It's not that we are *only* filtering for declined offers.  It's that we're only filtering for a "configured potentially longer period of time" for declined offers that have timed out in our RotatingMap.  Not sure how to phrase that nicely.  But it's *not* that we're *only* filtering for decline offers.
2. The mesos.proto docs and/or API docs should be improved a bit to make the purpose of `Filters.refuse_seconds` more clear.  i.e., "unused resources" is insufficiently explicated, as technically the offers being hoarded by MesosNimbus are also "unused", and thus I was getting confused thinking that this "refuse_seconds" functionality was related to rescinding/InverseOffers.  Perhaps if the comment simply said "refused resources" it would be clearer.  Or an explanation that "unused resources" *does* **not** include those offers that are being held by a framework and are as-of-yet unused.